From 8a919c9c6e84a1c3344cc0efc309582de26d3215 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 30 Apr 2019 18:28:27 +0300
Subject: [PATCH] [SW-4177] goya: uart/sf: calculate baud rate

Add dynamic baud rate calculation based on the settings in effect
(read from H/W) for both UART and the SPI flash.

Change-Id: I2ad72ded439854db44775c9b593914aec4b25852
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/goya/goya.c b/goya/goya.c
index 748f81e..af4e41f 100644
--- a/goya/goya.c
+++ b/goya/goya.c
@@ -14,6 +14,7 @@
 
 #include <preboot.h>
 #include <pll.h>
+#include <uart.h>
 
 #include "goya.h"
 
@@ -211,11 +212,21 @@
 	return pll_num;
 }
 
-void hl_uart_init(void) {};
+void hl_uart_init(void)
+{
+	u32 pcie_div2_freq = hl_pll_get_freq(mmPSOC_PCI_PLL_BASE, HL_PLL_DIV2);
+
+	preboot_uart_init(pcie_div2_freq / HL_UART_BAUDRATE);
+}
 
 u32 hl_spi_baudr_get(void)
 {
-	return 0x2;	/* Min divisor */
+	u32 factor;
+	u32 pcie_div2_freq = hl_pll_get_freq(mmPSOC_PCI_PLL_BASE, HL_PLL_DIV2);
+
+	factor = pcie_div2_freq / HL_SPI_FLASH_MAX_FREQ;
+
+	return factor > 2 ? factor : 2;
 }
 
 int hl_pll_partial_init_get(void)
diff --git a/goya/goya.h b/goya/goya.h
index 0b1b996..4e2256e 100644
--- a/goya/goya.h
+++ b/goya/goya.h
@@ -8,3 +8,4 @@
 #define mmCPU_PLL_BASE		0x4A2000
 #define mmIC_PLL_BASE		0x4A3000
 #define mmPSOC_MME_PLL_BASE	0xC71000
+#define mmPSOC_PCI_PLL_BASE	0xC72000
